### PR TITLE
[8.0][IMP] Transfer the Gross Profit fields to the Deliverable lines

### PR DESCRIPTION
**Pull request by sudhir-serpentcs**
Transfer the Gross Profit fields to the Deliverable lines.

BR with BRD_cost give the calculation for Total Revenue-Total Cost = Gross Profit
Need to have on every DL the same calculation based on their own RL
So DL 1 will have a partial Total Revenue DL 1 - Total Cost DL 1 = Gross Profit DL 1

@elicoidal

### CODE REVIEW

File: business_requirement_deliverable_cost/models/business.py [hunk 1]
```diff
@@ -123,6 +123,24 @@ class BusinessRequirementDeliverable(models.Model):
         groups='business_requirement_deliverable_cost.'
         'group_business_requirement_estimation',
     )
+    gross_profit = fields.Float(
+        compute='_compute_gross_profit',
+        string='Gross Profit',
+    )
+
+    @api.multi
+    @api.depends('resource_ids.price_total',
```
Comment (elicoidal) [inline business_requirement_deliverable_cost/models/business.py]: I am not sure this does what I am expecting.
Here is what I need: DL.gross_profit = DL.price_total - sum(RL.cost)
It should be updated whenever the RL.cost or DL.price_total changes.

Can you double check? @sudhir-serpentcs

File: business_requirement_deliverable_cost/views/business.xml [hunk 2]
```diff
@@ -31,6 +31,9 @@
                                 class="oe_highlight"/>
                     </group>
                 </xpath>
+                <xpath expr="//field[@name='price_total']" position="after">
+                    <field name="gross_profit"/>
```
Comment (elicoidal) [inline business_requirement_deliverable_cost/views/business.xml]: Not only: I would like to see the same fields as in the BR (check screenshot that I have provided).
Besides it should only be visible to the "cost" group

**Review by elicoidal — CHANGES_REQUESTED** [review]
Please review the logic.
Please review the form fields to be added
Please review the ACL.

**Comment by sudhir-serpentcs** [comment]
@elicoidal  
Changed calculation for gross profit and added ACL.
https://github.com/OCA/business-requirement/pull/105/commits/a59e162ee32249967115198f7003980f14f06225

**Comment by elicoidal** [comment]
@sudhir-serpentcs please check this and the screenshot. The current code doesnot cover it all AFAIK.

**Comment by elicoidal** [comment]
please follow: https://github.com/OCA/business-requirement/issues/84

**Comment by sudhir-serpentcs** [comment]
@elicoidal @victormartinelicocorp 
can you please review code.

**Review by elicoidal — APPROVED** [review]
Functional tests are OK

**Comment by elicoidal** [comment]
Will need to be fixed when #113 is finished

**Comment by elicoidal** [comment]
@sudhir-serpentcs can you rebase your work to fit the changes made in #113

**Comment by elicoidal** [comment]
@sudhir-serpentcs once you fix the link to #113 we could quickly merge

**Comment by elicoidal** [comment]
@sudhir-serpentcs #113 is fixed. Can you rebase and fix this one?

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
please review code.

**Comment by elicoidal** [comment]
@sudhir-serpentcs I have lost the description field in deliverable

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
Added Description field, please review code.

**Comment by elicoidal** [comment]
@sudhir-serpentcs far from been ready: can you retest the full process?

When creating a RL in the DL, I have following js error: Error: NameError: name 'business_requirement_id' is not defined

**Comment by sudhir-serpentcs** [comment]
@elicoidal  
Improved code as per comment.
please review code

**Comment by elicoidal** [comment]
@sudhir-serpentcs thanks it works now.
I want to change the position of the fields as it is a bit weird as is (so you will have probably to update both modules).

First column:
* Business Requirement
* Product
* Name

Second column:
* Quantity 
* Sales Price
* Currency
* Total revenue
* Total procurement
* Total tasks
* Estimated Gross Profit

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
changed the position of the fields.
please review code.

**Comment by elicoidal** [comment]
@sudhir-serpentcs Position of the fields is correct in DL form. Thanks!

I have the following small improvements:
- [X]  Total revenue+currency should appear in BR tree view and total at bottom (#108). Should appear in pivot view.
- [X] RL contains price/cost info but not when they are in DL.
- [X] Currency is not set if no customer/pricelist is select. It should be the company currency by default.
- [X] Responsible is not in the BR tree View. It should be the first
- [X] Remove "Created on" in the BR Tree view
- [X] the smart button Deliverable Lines should add the total revenue between parenthesis. Rename Deliverable Lines => Deliverable
- [X] the smart button Resource Lines should add the total cost between parenthesis. Rename Resources Lines => Resources

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
Changed code as per comment.
please review code.

**Comment by elicoidal** [comment]
@sudhir-serpentcs almost there. A couple of details left:
- [X] RL contains Sales price/Revenue info but not when they are in DL (my previous comment was not clear)
- [X] add the currency in tab Cost control in BR, same as in deliverable.
- [ ] In BR pivot: I dont have the Total revenue, Total Cost and Gross profit as measure
- [ ] In deliverable pivot: I dont have the Total revenue, Total Cost and Gross profit as measure
- [X] In resource pivot: I dont have the Total revenue and Total Cost as measure (While they appear in the table)

Rest looks awesome!

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
Improved pivot views & code as per comments.
please review code.

**Comment by elicoidal** [comment]
@sudhir-serpentcs still not there (check: https://github.com/OCA/business-requirement/pull/105#issuecomment-297589631):
- [ ] DL and BR pivot are not complete. 
- [ ] BR pivot should be table by default and is broken. Not all measure are available
- [ ] DL Pivot: I miss cost fields (unit price and total)

**Comment by sudhir-serpentcs** [comment]
@elicoidal 

>DL Pivot: I miss cost fields (unit price and total)

Changed code as per comments, except the above point as I could not find the mentioned fields.

Please review code.

CC @victormartinelicocorp @darshan-serpent

File: business_requirement/views/business_view.xml [hunk 3]
```diff
@@ -107,6 +107,7 @@
             <field name="model">business.requirement</field>
             <field name="arch" type="xml">
                 <tree string="Business Requirement">
+                    <field name="responsible_id"/>
```
Comment (dreispt) [inline business_requirement/views/business_view.xml]: Missing one indentation space

File: business_requirement/views/business_view.xml [hunk 4]
```diff
@@ -135,7 +135,7 @@
                         <filter string="Customer " icon="terp-go-home"  context="{'group_by':'partner_id'}"/>
                         <filter string="State" icon="terp-go-home" context="{'group_by':'state'}"/>
                         <filter string="Priority" icon="terp-go-home" context="{'group_by':'priority'}"/>
-                        <filter string="Responsible" icon="terp-go-home" context="{'group_by':'required'}"/>
+                        <filter string="Responsible" icon="terp-go-home" context="{'group_by':'responsible_id'}"/>
```
Comment (dreispt) [inline business_requirement/views/business_view.xml]: Icons are deprecated afaik

File: business_requirement_deliverable/models/business.py [hunk 5]
```diff
@@ -144,12 +144,13 @@ class BusinessRequirementDeliverable(models.Model):
...
         compute='_compute_get_price_total',
         string='Total revenue',
+        store=True
     )
     currency_id = fields.Many2one(
         comodel_name='res.currency',
         string='Currency',
         readonly=True,
-        compute='_compute_get_currency',
+        compute='_compute_get_currency'
```
Comment (dreispt) [inline business_requirement_deliverable/models/business.py]: No problem with leaving the trailing comma here

File: business_requirement_deliverable/models/business.py [hunk 6]
```diff
@@ -172,6 +173,8 @@ def _compute_get_currency(self):
             currency_id = partner_id.property_product_pricelist.currency_id
             if currency_id:
                 brd.currency_id = currency_id
+            else:
+                brd.currency_id = self.env.user.company_id.currency_id.id
```
Comment (dreispt) [inline business_requirement_deliverable/models/business.py]: Afaict you should remove the trailing `.id`: you want to reference the Currency record, and not it's numeric id.

File: business_requirement_deliverable/tests/test_br.py [hunk 7]
```diff
@@ -149,7 +149,13 @@ def test_open_business_requirement_dl_rl(self):
         self.br.open_deliverable_line()
         self.br.open_resource_line()
 
+    def test_compute_total_revenue_cost(self):
+        self.br._compute_dl_total_revenue()
+        self.br._compute_rl_total_cost()
```
Comment (dreispt) [inline business_requirement_deliverable/tests/test_br.py]: You're not actually testing the code here.
You should add lines to a BR, and then read the totals and compare with the expected value...

File: business_requirement_deliverable_cost/models/business.py [hunk 8]
```diff
@@ -181,7 +227,7 @@ def _compute_resource_task_total(self):
                     br.mapped('deliverable_lines').mapped(
                         'resource_ids').filtered(
                         lambda r: r.resource_type == 'task').mapped(
-                            'price_total'))
+                        'price_total'))
```
Comment (dreispt) [inline business_requirement_deliverable_cost/models/business.py]: The indentation was correct.

File: business_requirement_deliverable_cost/models/business.py [hunk 9]
```diff
@@ -192,7 +238,7 @@ def _compute_resource_procurement_total(self):
                     br.mapped('deliverable_lines').mapped(
                         'resource_ids').filtered(
                         lambda r: r.resource_type == 'procurement').mapped(
-                            'price_total'))
+                        'price_total'))
```
Comment (dreispt) [inline business_requirement_deliverable_cost/models/business.py]: Don't remove the indentation.

**Review by dreispt — COMMENTED** [review]
LGTM, but a few nitpicks to address

**Comment by elicoidal** [comment]
Pivot view is working :+1: 

> DL Pivot: I miss cost fields (unit price and total)
Changed code as per comments, except the above point as I could not find the mentioned fields.

You are right they must be function (maybe make them storable in both BR and DL): Total procurement / Total tasks

**Comment by elicoidal** [comment]
thanks @dreispt for your review

**Comment by sudhir-serpentcs** [comment]
@dreispt Have made the requested changes.

@elicoidal And have added Total procurement / Total tasks to the pivot view, yes had to make it storable.

thanks for your review guys!

**Review by dreispt — APPROVED** [review]
LGTM. Thanks for the patience.

**Comment by elicoidal** [comment]
LGTM too! merging

**Pull request merged**